packages/web/services/twitter.ts

```diff
@@ -0,0 +1,112 @@
+  createdAt: string;
+  user: {
+    username: string;
+    name: string;
+    profilePictureUrl: string | null;
+    url: string | null;
+  };
+  previewImage: string | null;
+}
+
+export interface TwitterServiceInterface {
+  getUserTweets: (userId: string) => Promise<RichTweet[]>;
+}
+
+const TwitterInternal: TwitterServiceInterface = {
+  /**
+   * Make a request to Twitter services to search for tweets based on a userId.
+   *
+   * By default it returns the most recent ten tweets https://developer.twitter.com/en/docs/twitter-api/tweets/timelines/api-reference/get-users-id-tweets
+   *
+   * @param userId a search string (e.g. "osmosiszone")
+   * @returns An array of tweet's objects
+   */
+  getUserTweets: async (userId: string) => {
+    const {
+      data: {
+        data: tweets,
+        includes: { users, media },
+      },
+    } = await twitterApi.get(
+      `/tweets/search/recent?query=${encodeURIComponent(
+        `from:${userId}`
+      )}&max_results=10&tweet.fields=created_at&expansions=author_id,attachments.media_keys&media.fields=media_key,type,url&user.fields=description,profile_image_url,url`
+    );
+
+    return tweets.map((tweet: any) => {
+      const userTweet = users.find((u: any) => u.id === tweet.author_id);
+      let mediaTweet: any | undefined;
+      tweet.attachments?.media_keys.forEach((media_key: any) => {
+        mediaTweet = media.find(
+          (m: any) => m.media_key === media_key && m.type === "photo"
+        );
+        if (mediaTweet) {
+          return;
+        }
+      });
+
+      return {
+        id: tweet.id,
+        text: tweet.text,
+        createdAt: tweet.created_at,
+        user: {
+          username: userTweet.username,
+          name: userTweet.name,
+          profilePictureUrl: userTweet.profile_image_url ?? null,
+          url: userTweet.url ?? null,
+        },
+        previewImage: mediaTweet !== undefined ? mediaTweet.url : null,
+      };
+    });
+  },
+};
+
+/**
+ * Expire time in milliseconds. (7 days)
+ *
+ * It should be enought if we wanna get tweets from 35 tokens
+ * every 7 days.
+ */
+const CACHE_EXPIRE_TIME = 1000 * 60 * 60 * 24 * 7;
+
+export const Twitter: TwitterServiceInterface = {
+  /**
+   * Returns the tweets associated with the userId passed as a parameter,
+   * if it finds them in cache it returns them if not it makes a request
+   * to Twitter and then caches the result
+   *
+   * @param userId a search string (e.g. "osmosiszone")
+   * @returns An array of tweet's objects
+   */
+  getUserTweets: async (userId: string) => {
+    const twitter = createClient({
+      url: process.env.KV_REST_API_URL!,
+      token: process.env.KV_REST_API_TOKEN!,
+    });
+
+    let cachedTweets = await twitter.get<RichTweet[]>(userId);
+
+    if (!cachedTweets) {
+      cachedTweets = await TwitterInternal.getUserTweets(userId);
+
+      twitter.set(userId, cachedTweets, { px: CACHE_EXPIRE_TIME, nx: true });
+    }
+
+    return cachedTweets;
+  },
+};
```

Comment:
I suggest creating a class to streamline the underlying logic. Additionally, I recommend utilizing the `cachified` utility from our [queries](https://github.com/osmosis-labs/osmosis-frontend/blob/85b7c7372734044139ff4d09f9677d38a491e35c/packages/web/queries/complex/pools.ts#L92). It's straightforward to implement and manages all caching, even with Vercel KV using the Redis adapter."

Reply:
Done here: https://github.com/osmosis-labs/osmosis-frontend/pull/2234/commits/458a15d69724e86de92e0962e873612b01bcad03